uber/tests/models/test_badge_funcs.py

```diff
@@ -206,6 +310,80 @@ def test_non_preassigned_deletion(self, session):
         session.delete(session.query(Attendee).filter_by(badge_type=c.ATTENDEE_BADGE).first())
         session.commit()
 
-# TODO: more unit tests for changes/deletions after custom badges have been ordered
+    def test_delete_shift_not_enabled(self, session, monkeypatch):
+        monkeypatch.setattr(c, 'SHIFT_CUSTOM_BADGES', False)
+        session.delete(session.staff_one)
+        session.commit()
+
+    def test_delete_custom_badges_ordered(self, session, monkeypatch):
+        monkeypatch.setattr(c, 'PRINTED_BADGE_DEADLINE', datetime.now(UTC)-timedelta(days=1))
+        session.delete(session.staff_one)
+        session.commit()
+
+
+class TestShiftOnChange:
+    def staff_badges(self, session):
+        return sorted(a.badge_num for a in session.query(Attendee).filter(Attendee.badge_status != c.INVALID_STATUS).filter_by(badge_type=c.STAFF_BADGE).all())
+
+    def test_shift_on_add(self, session):
+        assert 'Badge updated' == session.update_badge(Attendee(first_name='NewStaff', paid=c.NEED_NOT_PAY, badge_type=c.STAFF_BADGE, badge_num=2), None, None)
+        assert session.staff_two.badge_num == 3
+        assert [1, 3, 4, 5, 6] == self.staff_badges(session)
```

Comment:
Is the reason why this is ``[1, 3, 4, 5, 6]`` with a missing ``2`` because the new Attendee is ends up with an invalid status?